### PR TITLE
dodge path re-encoding issues on Windows in list.files() and friends

### PR DESCRIPTION
**Pull request by kevinushey**
### Intent

Addresses https://github.com/rstudio/rstudio/issues/10451.

### Approach

The fix here is somewhat heavy-handed: we overwrite R's implementation of `list.files()`, `list.dirs()`, and `dir()`, with our own versions that better handle different combinations of {system encoding} vs {path encoding}.

This is necessary because RStudio now declares itself to be compatible with the UTF-8 codepage 65001, and so paths returned from R's `list.files()` are actually UTF-8 encoded. However, R does not mark the returned files as such, and so they're mis-interpreted in the native encoding, which may not be UTF-8 for older versions of R.

We still make use of the same underlying internal `list.files()` implementation; we just massage the file encodings + how these functions are invoked in a way that makes sure they behave regardless of the current locale.

### Automated Tests

Tests included, but since they're not run on Windows we'll need extra QA testing.

### QA Notes

Test via notes in https://github.com/rstudio/rstudio/issues/10451. This _definitely_ needs extra scrutiny since we're messing with a base R function!

### Checklist

- [x] If this PR adds a new feature, or fixes a bug in a previously released version, it includes an entry in `NEWS.md` 
- [x] If this PR adds or changes UI, the updated UI meets [accessibility standards](https://github.com/rstudio/rstudio/wiki/Accessibility)
- [x] A reviewer is assigned to this PR (if unsure who to assign, check Area Owners list)
- [x] This PR passes all local unit tests

### CODE REVIEW

**Comment by MariaSemple** [comment]
@kevinushey did you intend to merge this into the PT release? If so, can you rebase it to `prairie-trillium`?

**Comment by jgutman** [comment]
Since Kevin mentions that this requires "extra QA testing" and "extra scrutiny" maybe it's best to keep it in main for Spotted Wakerobin?

**Comment by kevinushey** [comment]
I was also wondering whether we should just provide our own implementation (e.g. built on top of Boost.Filesystem) that tries to mimic R's own implementations as closely as possible.

File: src/cpp/session/modules/SessionFiles.cpp [hunk 1]
```diff
@@ -1397,8 +1409,9 @@ Error initialize()
    // subscribe to events
    events().onClientInit.connect(bind(onClientInit));
 
-   RS_REGISTER_CALL_METHOD(rs_readLines, 1);
-   RS_REGISTER_CALL_METHOD(rs_pathInfo, 1);
+   RS_REGISTER_CALL_METHOD(rs_getACP);
```
Comment (jgutman) [inline src/cpp/session/modules/SessionFiles.cpp:1412]: Is there a reason for removing the 1 from the RS_REGISTER_CALL_METHOD calls?

**Review by jgutman — APPROVED** [review]
LGTM, although I'm curious to know what alternative approaches are possible.

Also, can you clarify, when we bind our functions to these base R functions, what does the user see if they try to look at the source code for the original function with `View()` or the help page with `?` (not that I'd expect the source code to be particularly helpful since it's just `.Internal`)

File: src/cpp/session/modules/SessionFiles.cpp (hunk 1 above)
Comment (kevinushey) [reply src/cpp/session/modules/SessionFiles.cpp:1412]: Those are sort of unneeded historical artifacts. Back when we were compiling against C++98, we didn't have a clean way of figuring out how many parameters each call method received, and so we have to explicitly declare it. That was a bit error-prone, as we occasionally forgot to update our registration code whenever we updated / extended an existing `rs_` method.

Now that we can figure it out using C++11 tooling (https://github.com/rstudio/rstudio/blob/89d90af57611a1d5e42b3dfdd2abf3565d029ea1/src/cpp/r/include/r/RRoutines.hpp#L80) I've been slowly cleaning up some of those usages as I see them.

**Comment by kevinushey** [comment]
> LGTM, although I'm curious to know what alternative approaches are possible.

The other thing we could do is try to implement our own "list files" functionality, that tries to behave as similarly as R's own internal implementation as possible. I wasn't sure if we'd be able to match all of the nuances of R's implementation ourselves, so I tried this approach first (which has its own caveats; e.g. I'm not sure if we can be sure we can always `setwd()`

> Also, can you clarify, when we bind our functions to these base R functions, what does the user see if they try to look at the source code for the original function with View() or the help page with ? (not that I'd expect the source code to be particularly helpful since it's just .Internal)

In this case, they'd see our implementation text here: https://github.com/rstudio/rstudio/pull/10464/files?w=1#diff-c6b73983b11a05d33626cb51249e44dfa5c6171f65975ba52439cf4a9c379abaR39-R51

**Comment by kevinushey** [comment]
Superceded by https://github.com/rstudio/rstudio/pull/10481.

**Pull request closed without merging**